File: send_report.js
import "dotenv/config";
import nodemailer from "nodemailer";

const TODAY = new Date().toLocaleDateString("en-CA", {
  timeZone: "America/New_York",
});

const RECIPIENT = "andrewric86@gmail.com";

// ── 1. Fetch live market data ─────────────────────────────────────────────────

async function fetchMarketData() {
  const [cgRes, fgRes] = await Promise.all([
    fetch(
      "https://api.coingecko.com/api/v3/coins/bitcoin?localization=false&tickers=false&community_data=false&developer_data=false"
    ),
    fetch("https://api.alternative.me/fng/?limit=2"),
  ]);

  const cg = await cgRes.json();
  const fg = (await fgRes.json()).data;
  const md = cg.market_data;

  return {
    price:        md.current_price.usd,
    change24h:    md.price_change_percentage_24h,
    change7d:     md.price_change_percentage_7d,
    change30d:    md.price_change_percentage_30d,
    volume24h:    md.total_volume.usd,
    marketCap:    md.market_cap.usd,
    ath:          md.ath.usd,
    athChangePct: md.ath_change_percentage.usd,
    high24h:      md.high_24h.usd,
    low24h:       md.low_24h.usd,
    fgValue:      parseInt(fg[0].value),
    fgLabel:      fg[0].value_classification,
    fgPrev:       parseInt(fg[1].value),
  };
}

// ── 2. Rule-based analysis ────────────────────────────────────────────────────

function analyse(d) {
  const { price, change24h, change7d, change30d, volume24h, marketCap, ath, athChangePct, high24h, low24h, fgValue, fgPrev } = d;

  // Direction
  const direction =
    change24h > 2 && change7d > 3 ? "UP" :
    change24h < -2 && change7d < -3 ? "DOWN" : "SIDEWAYS";

  // Volume
  const volRatio = (volume24h / marketCap) * 100;
  const volume = volRatio > 6 ? "HIGH" : volRatio < 2.5 ? "LOW" : "NORMAL";

  // Sentiment
  const sentiment =
    fgValue >= 75 ? { label: "EXTREME GREED", note: "Market is overheated — risk of sharp pullback is elevated." } :
    fgValue >= 55 ? { label: "GREED",         note: "Bullish momentum present but watch for complacency." } :
    fgValue >= 45 ? { label: "NEUTRAL",        note: "No strong directional bias — wait for a clearer signal." } :
    fgValue >= 25 ? { label: "FEAR",           note: "Bearish sentiment dominates — historically a buying opportunity area." } :
                    { label: "EXTREME FEAR",   note: "Maximum pessimism — often marks cycle lows for patient buyers." };

  // Stance
  const stance =
    fgValue >= 55 && change7d > 0 && change30d > 0 ? "BULLISH" :
    fgValue <= 40 && change7d < 0 && change30d < 0 ? "BEARISH" : "NEUTRAL";

  // Key levels
  const support1     = Math.round(low24h  * 0.99 / 100)  * 100;
  const support2     = Math.round(price   * 0.95 / 500)   * 500;
  const resist1      = Math.round(high24h * 1.01 / 100)   * 100;
  const resist2      = Math.round(price   * 1.05 / 500)   * 500;
  const invalidation = Math.round(price   * 0.92 / 500)   * 500;

  // ATH note
  const athPct = Math.abs(athChangePct);
  const athNote =
    athPct < 5  ? `only ${athPct.toFixed(1)}% below ATH ($${fmt(ath)}) — price discovery territory` :
    athPct < 15 ? `${athPct.toFixed(1)}% below ATH ($${fmt(ath)}) — upper range of historical cycle` :
                  `${athPct.toFixed(1)}% below ATH ($${fmt(ath)}) — mid-cycle range`;

  // Narratives
  const shortTerm =
    direction === "UP"   ? `Positive momentum; buyers in control above $${fmt(support1)}` :
    direction === "DOWN" ? `Selling pressure active; watch $${fmt(support1)} as near-term floor` :
                           `Coiling between $${fmt(support1)} and $${fmt(resist1)} — breakout imminent`;

  const mediumTerm =
    direction === "UP"   ? `Bullish structure intact if price holds above $${fmt(support2)}` :
    direction === "DOWN" ? `Recovery needs reclaim of $${fmt(resist1)} to shift bias` :
                           `Trend continuation depends on which level breaks first`;

  // Scenarios
  const bull =
    stance === "BULLISH" ? `Hold above $${fmt(resist1)} → continuation toward $${fmt(resist2)}+` :
    stance === "BEARISH" ? `Reclaim of $${fmt(resist1)} and hold → relief rally to $${fmt(resist2)}` :
                           `Break above $${fmt(resist1)} with volume → push toward $${fmt(resist2)}`;

  const bear =
    stance === "BULLISH" ? `Rejection + close below $${fmt(support1)} → retest of $${fmt(support2)}` :
    stance === "BEARISH" ? `Continued selling below $${fmt(support1)} → test of $${fmt(support2)}` :
                           `Break below $${fmt(support1)} → accelerated move to $${fmt(support2)}`;

  const base =
    stance === "BULLISH" ? `Consolidation near $${fmt(price)}, then upside continuation (higher probability)` :
    stance === "BEARISH" ? `Further downside toward $${fmt(support2)} before stabilisation` :
                           `Range-bound between $${fmt(support1)}–$${fmt(resist1)} pending catalyst`;

  // Confidence
  const signals = [
    Math.abs(change24h) > 2,
    Math.abs(change7d)  > 5,
    fgValue < 25 || fgValue > 75,
    volRatio > 5 || volRatio < 2,
  ].filter(Boolean).length;

  const confidence =
    signals >= 3 ? "HIGH" :
    signals === 2 ? "MEDIUM-HIGH" :
    signals === 1 ? "MEDIUM" : "LOW";

  // News drivers
  const drivers = [];
  if (Math.abs(change24h) > 3) {
    const move = change24h > 0 ? "rallied" : "sold off";
    drivers.push(`<strong>Strong 24h move:</strong> BTC ${move} ${Math.abs(change24h).toFixed(1)}% — significant one-day volatility.`);
  }
  if (athPct < 10) drivers.push(`<strong>Near all-time high:</strong> Price is ${athPct.toFixed(1)}% below ATH ($${fmt(ath)}) — resistance and profit-taking risk elevated.`);
  if (fgValue <= 25) drivers.push(`<strong>Extreme Fear (${fgValue}):</strong> Historically these levels coincide with capitulation bottoms — potential contrarian buy zone.`);
  else if (fgValue >= 75) drivers.push(`<strong>Extreme Greed (${fgValue}):</strong> Crowd is very optimistic — corrections often follow these readings.`);
  if (change30d > 20)  drivers.push(`<strong>Strong monthly run:</strong> BTC is up ${change30d.toFixed(1)}% over 30 days — momentum strong but extended.`);
  else if (change30d < -15) drivers.push(`<strong>Monthly drawdown:</strong> BTC is down ${Math.abs(change30d).toFixed(1)}% over 30 days — accumulation zone for long-term holders.`);
  if (volRatio > 6) drivers.push(`<strong>Elevated volume:</strong> 24h volume is ${volRatio.toFixed(1)}% of market cap — unusual activity, potential breakout or breakdown.`);
  drivers.push(`<strong>ATH context:</strong> Current price is ${athNote}.`);

  return {
    direction, volume, volRatio, stance,
    sentiment, shortTerm, mediumTerm,
    bull, bear, base,
    confidence, signals,
    support1, support2, resist1, resist2, invalidation,
    drivers: drivers.slice(0, 5),
  };
}

// ── 3. Build HTML ─────────────────────────────────────────────────────────────

function buildHtml(d, a) {
  const sign = n => n >= 0 ? "+" : "";
  const fgArrow = d.fgValue > d.fgPrev ? "▲" : d.fgValue < d.fgPrev ? "▼" : "—";

  return `<h2>Bitcount (BTC) Daily Market Brief – ${TODAY}</h2>

<h3>Summary</h3>
<p>
  Bitcoin is trading at <strong>$${fmt(d.price)}</strong>
  (${sign(d.change24h)}${d.change24h.toFixed(2)}% in 24h,
  ${sign(d.change7d)}${d.change7d.toFixed(1)}% over 7 days).
  Trend is <strong>${a.direction}</strong> with <strong>${a.volume}</strong> volume
  and a <strong>${a.stance}</strong> market bias.
  Fear &amp; Greed: <strong>${d.fgValue} (${d.fgLabel})</strong> ${fgArrow} from ${d.fgPrev} yesterday.
</p>

<h3>News Drivers</h3>
<ul>
  ${a.drivers.map(item => `<li>${item}</li>`).join("\n  ")}
</ul>

<h3>Trend Analysis</h3>
<ul>
  <li><strong>Direction:</strong> ${a.direction}</li>
  <li><strong>Volume:</strong> ${a.volume} (${a.volRatio.toFixed(1)}% volume-to-market-cap)</li>
  <li><strong>Short-term (1–3d):</strong> ${a.shortTerm}</li>
  <li><strong>Medium-term (1–2w):</strong> ${a.mediumTerm}</li>
  <li><strong>24h range:</strong> $${fmt(d.low24h)} – $${fmt(d.high24h)}</li>
  <li><strong>30d change:</strong> ${sign(d.change30d)}${d.change30d.toFixed(1)}%</li>
</ul>

<h3>Sentiment</h3>
<ul>
  <li><strong>Overall:</strong> ${a.stance}</li>
  <li><strong>Fear &amp; Greed:</strong> ${d.fgValue} (${d.fgLabel}) ${fgArrow} vs ${d.fgPrev} yesterday</li>
  <li><strong>Context:</strong> ${a.sentiment.note}</li>
</ul>

<h3>Next Move Scenarios</h3>
<ul>
  <li><strong>Bull case:</strong> ${a.bull}</li>
  <li><strong>Bear case:</strong> ${a.bear}</li>
  <li><strong>Base case:</strong> ${a.base}</li>
</ul>

<h3>Key Levels</h3>
<ul>
  <li><strong>Support:</strong> $${fmt(a.support1)} / $${fmt(a.support2)}</li>
  <li><strong>Resistance:</strong> $${fmt(a.resist1)} / $${fmt(a.resist2)}</li>
  <li><strong>Invalidation:</strong> $${fmt(a.invalidation)}</li>
</ul>

<h3>Confidence</h3>
<p><strong>${a.confidence}</strong> — ${a.signals}/4 confirming signals (24h move, 7d trend, sentiment extreme, volume).</p>

<hr>
<p><em>Generated by bitcount-market-scout · ${TODAY} 07:00 ET</em></p>`;
}

// ── 4. Send email ─────────────────────────────────────────────────────────────

async function sendEmail(html) {
  const transporter = nodemailer.createTransport({
    service: "gmail",
    auth: {
      user: process.env.GMAIL_USER,
      pass: process.env.GMAIL_APP_PASSWORD,
    },
  });

  await transporter.sendMail({
    from: process.env.GMAIL_USER,
    to: RECIPIENT,
    subject: `Bitcount daily market brief – ${TODAY}`,
    text: html.replace(/<[^>]+>/g, "").replace(/\n{3,}/g, "\n\n").trim(),
    html,
  });

  console.log(`Sent: Bitcount daily market brief – ${TODAY}`);
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function fmt(n) {
  return n.toLocaleString("en-US");
}

// ── Main ──────────────────────────────────────────────────────────────────────

const data = await fetchMarketData();
console.log(`BTC: $${fmt(data.price)}  |  24h: ${data.change24h >= 0 ? "+" : ""}${data.change24h.toFixed(2)}%  |  F&G: ${data.fgValue} (${data.fgLabel})`);

const analysis = analyse(data);
const html = buildHtml(data, analysis);
await sendEmail(html);

File: send_report.py
#!/usr/bin/env python3
"""
Bitcount (BTC) daily market report mailer.
Fetches live data from CoinGecko + Fear & Greed API,
generates a structured report with rule-based analysis,
then emails it to the recipient.
No LLM API required.
"""

import smtplib
import os
import requests
from datetime import datetime
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from zoneinfo import ZoneInfo

ET = ZoneInfo("America/New_York")
TODAY = datetime.now(ET).strftime("%Y-%m-%d")

RECIPIENT    = "andrewric86@gmail.com"
GMAIL_USER   = os.environ["GMAIL_USER"]
GMAIL_APP_PW = os.environ["GMAIL_APP_PASSWORD"]

SUBJECT = f"Bitcount daily market brief – {TODAY}"


# ── 1. Fetch live market data ─────────────────────────────────────────────────

def fetch_market_data() -> dict:
    cg = requests.get(
        "https://api.coingecko.com/api/v3/coins/bitcoin",
        params={"localization": "false", "tickers": "false",
                "community_data": "false", "developer_data": "false"},
        timeout=15,
    ).json()

    md = cg["market_data"]
    data = {
        "price":          md["current_price"]["usd"],
        "change_24h":     md["price_change_percentage_24h"],
        "change_7d":      md["price_change_percentage_7d"],
        "change_30d":     md["price_change_percentage_30d"],
        "volume_24h":     md["total_volume"]["usd"],
        "market_cap":     md["market_cap"]["usd"],
        "ath":            md["ath"]["usd"],
        "ath_change_pct": md["ath_change_percentage"]["usd"],
        "high_24h":       md["high_24h"]["usd"],
        "low_24h":        md["low_24h"]["usd"],
        "circulating":    md["circulating_supply"],
    }

    fg = requests.get(
        "https://api.alternative.me/fng/?limit=2", timeout=10
    ).json()["data"]
    data["fg_value"] = int(fg[0]["value"])
    data["fg_label"] = fg[0]["value_classification"]
    data["fg_prev"]  = int(fg[1]["value"])

    return data


# ── 2. Rule-based analysis ────────────────────────────────────────────────────

def analyse(d: dict) -> dict:
    p    = d["price"]
    c24  = d["change_24h"]
    c7   = d["change_7d"]
    c30  = d["change_30d"]
    fg   = d["fg_value"]
    ath  = d["ath"]
    vol  = d["volume_24h"]
    mc   = d["market_cap"]

    # Direction
    if c24 > 2 and c7 > 3:
        direction = "UP"
    elif c24 < -2 and c7 < -3:
        direction = "DOWN"
    else:
        direction = "SIDEWAYS"

    # Volume (volume-to-market-cap ratio typical range ~2-6%)
    vol_ratio = (vol / mc) * 100
    if vol_ratio > 6:
        volume = "HIGH"
    elif vol_ratio < 2.5:
        volume = "LOW"
    else:
        volume = "NORMAL"

    # Sentiment label
    if fg >= 75:
        sentiment = "EXTREME GREED"
        sentiment_note = "Market is overheated — risk of sharp pullback is elevated."
    elif fg >= 55:
        sentiment = "GREED"
        sentiment_note = "Bullish momentum present but watch for complacency."
    elif fg >= 45:
        sentiment = "NEUTRAL"
        sentiment_note = "No strong directional bias; wait for a clearer signal."
    elif fg >= 25:
        sentiment = "FEAR"
        sentiment_note = "Bearish sentiment dominates — historically a buying opportunity area."
    else:
        sentiment = "EXTREME FEAR"
        sentiment_note = "Maximum pessimism — often marks cycle lows for patient buyers."

    # Overall stance
    if fg >= 55 and c7 > 0 and c30 > 0:
        stance = "BULLISH"
    elif fg <= 40 and c7 < 0 and c30 < 0:
        stance = "BEARISH"
    else:
        stance = "NEUTRAL"

    # Key levels (derived from current price and 24h range)
    support1  = round(d["low_24h"] * 0.99 / 100) * 100
    support2  = round(p * 0.95 / 500) * 500
    resist1   = round(d["high_24h"] * 1.01 / 100) * 100
    resist2   = round(p * 1.05 / 500) * 500
    invalidation = round(p * 0.92 / 500) * 500

    # ATH proximity
    ath_pct = abs(d["ath_change_pct"])
    if ath_pct < 5:
        ath_note = f"only {ath_pct:.1f}% below ATH (${ath:,.0f}) — price discovery territory"
    elif ath_pct < 15:
        ath_note = f"{ath_pct:.1f}% below ATH (${ath:,.0f}) — upper range of historical cycle"
    else:
        ath_note = f"{ath_pct:.1f}% below ATH (${ath:,.0f}) — mid-cycle range"

    # Short / medium-term narrative
    if direction == "UP":
        short_term  = f"Positive momentum; buyers in control above ${support1:,.0f}"
        medium_term = f"Bullish structure intact if price holds above ${support2:,.0f}"
    elif direction == "DOWN":
        short_term  = f"Selling pressure active; watch ${support1:,.0f} as near-term floor"
        medium_term = f"Recovery needs reclaim of ${resist1:,.0f} to shift bias"
    else:
        short_term  = f"Coiling between ${support1:,.0f} and ${resist1:,.0f} — breakout imminent"
        medium_term = f"Trend continuation depends on which level breaks first"

    # Bull / bear / base scenarios
    if stance == "BULLISH":
        bull = f"Hold above ${resist1:,.0f} → continuation toward ${resist2:,.0f}+"
        bear = f"Rejection + close below ${support1:,.0f} → retest of ${support2:,.0f}"
        base = f"Consolidation near ${p:,.0f}, then upside continuation (higher probability)"
    elif stance == "BEARISH":
        bull = f"Reclaim of ${resist1:,.0f} and hold → relief rally to ${resist2:,.0f}"
        bear = f"Continued selling below ${support1:,.0f} → test of ${support2:,.0f}"
        base = f"Further downside pressure toward ${support2:,.0f} before stabilisation"
    else:
        bull = f"Break above ${resist1:,.0f} with volume → push toward ${resist2:,.0f}"
        bear = f"Break below ${support1:,.0f} → accelerated move to ${support2:,.0f}"
        base = f"Range-bound between ${support1:,.0f}–${resist1:,.0f} pending catalyst"

    # Confidence
    signals = sum([
        abs(c24) > 2,
        abs(c7) > 5,
        fg < 25 or fg > 75,
        vol_ratio > 5 or vol_ratio < 2,
    ])
    if signals >= 3:
        confidence = "HIGH"
    elif signals == 2:
        confidence = "MEDIUM-HIGH"
    elif signals == 1:
        confidence = "MEDIUM"
    else:
        confidence = "LOW"

    conf_note = f"{signals}/4 confirming signals (24h move, 7d trend, sentiment extreme, volume)"

    # News drivers (derived from data)
    drivers = []
    if abs(c24) > 3:
        move = "rallied" if c24 > 0 else "sold off"
        drivers.append(f"<strong>Strong 24h move:</strong> BTC {move} {abs(c24):.1f}% in the last 24 hours — significant one-day volatility.")
    if ath_pct < 10:
        drivers.append(f"<strong>Near all-time high:</strong> Price is {ath_pct:.1f}% below ATH (${ath:,.0f}) — resistance and profit-taking risk elevated.")
    if fg <= 25:
        drivers.append(f"<strong>Extreme Fear ({fg}):</strong> Historically these levels coincide with capitulation bottoms — potential contrarian buy zone.")
    elif fg >= 75:
        drivers.append(f"<strong>Extreme Greed ({fg}):</strong> Crowd is very optimistic — corrections often follow these readings.")
    if c30 > 20:
        drivers.append(f"<strong>Strong monthly run:</strong> BTC is up {c30:.1f}% over 30 days — momentum strong but extended.")
    elif c30 < -15:
        drivers.append(f"<strong>Monthly drawdown:</strong> BTC is down {abs(c30):.1f}% over 30 days — accumulation zone for long-term holders.")
    if vol_ratio > 6:
        drivers.append(f"<strong>Elevated volume:</strong> 24h volume is {vol_ratio:.1f}% of market cap — unusual activity, potential breakout or breakdown.")
    # Always include ATH context
    drivers.append(f"<strong>ATH context:</strong> Current price is {ath_note}.")
    drivers = drivers[:5]  # cap at 5

    return {
        "direction": direction,
        "volume": volume,
        "vol_ratio": vol_ratio,
        "sentiment": sentiment,
        "sentiment_note": sentiment_note,
        "stance": stance,
        "support1": support1,
        "support2": support2,
        "resist1": resist1,
        "resist2": resist2,
        "invalidation": invalidation,
        "short_term": short_term,
        "medium_term": medium_term,
        "bull": bull,
        "bear": bear,
        "base": base,
        "confidence": confidence,
        "conf_note": conf_note,
        "drivers": drivers,
        "ath_note": ath_note,
    }


# ── 3. Build HTML report ──────────────────────────────────────────────────────

def build_html(d: dict, a: dict) -> str:
    drivers_html = "\n".join(f"  <li>{item}</li>" for item in a["drivers"])

    fg_arrow = "▲" if d["fg_value"] > d["fg_prev"] else ("▼" if d["fg_value"] < d["fg_prev"] else "—")

    return f"""<h2>Bitcount (BTC) Daily Market Brief – {TODAY}</h2>

<h3>Summary</h3>
<p>
  Bitcoin is trading at <strong>${d['price']:,.0f}</strong>
  ({'+' if d['change_24h'] >= 0 else ''}{d['change_24h']:.2f}% in 24h,
  {'+' if d['change_7d'] >= 0 else ''}{d['change_7d']:.1f}% over 7 days).
  The overall trend is <strong>{a['direction']}</strong> with <strong>{a['volume']}</strong> volume
  and a <strong>{a['stance']}</strong> market bias.
  The Fear &amp; Greed Index sits at <strong>{d['fg_value']} ({d['fg_label']})</strong> {fg_arrow} from {d['fg_prev']} yesterday.
</p>

<h3>News Drivers</h3>
<ul>
{drivers_html}
</ul>

<h3>Trend Analysis</h3>
<ul>
  <li><strong>Direction:</strong> {a['direction']}</li>
  <li><strong>Volume:</strong> {a['volume']} ({a['vol_ratio']:.1f}% volume-to-market-cap ratio)</li>
  <li><strong>Short-term (1–3d):</strong> {a['short_term']}</li>
  <li><strong>Medium-term (1–2w):</strong> {a['medium_term']}</li>
  <li><strong>24h range:</strong> ${d['low_24h']:,.0f} – ${d['high_24h']:,.0f}</li>
  <li><strong>30d change:</strong> {'+' if d['change_30d'] >= 0 else ''}{d['change_30d']:.1f}%</li>
</ul>

<h3>Sentiment</h3>
<ul>
  <li><strong>Overall:</strong> {a['stance']}</li>
  <li><strong>Fear &amp; Greed Index:</strong> {d['fg_value']} ({d['fg_label']}) {fg_arrow} vs {d['fg_prev']} yesterday</li>
  <li><strong>Context:</strong> {a['sentiment_note']}</li>
</ul>

<h3>Next Move Scenarios</h3>
<ul>
  <li><strong>Bull case:</strong> {a['bull']}</li>
  <li><strong>Bear case:</strong> {a['bear']}</li>
  <li><strong>Base case:</strong> {a['base']}</li>
</ul>

<h3>Key Levels</h3>
<ul>
  <li><strong>Support:</strong> ${a['support1']:,.0f} / ${a['support2']:,.0f}</li>
  <li><strong>Resistance:</strong> ${a['resist1']:,.0f} / ${a['resist2']:,.0f}</li>
  <li><strong>Invalidation:</strong> ${a['invalidation']:,.0f}</li>
</ul>

<h3>Confidence</h3>
<p><strong>{a['confidence']}</strong> — {a['conf_note']}.</p>

<hr>
<p><em>Generated by bitcount-market-scout · {TODAY} 07:00 ET</em></p>"""


# ── 4. Send email ─────────────────────────────────────────────────────────────

def send(html_body: str):
    import re
    plain = re.sub(r"<[^>]+>", "", html_body).strip()

    msg = MIMEMultipart("alternative")
    msg["Subject"] = SUBJECT
    msg["From"]    = GMAIL_USER
    msg["To"]      = RECIPIENT

    msg.attach(MIMEText(plain, "plain"))
    msg.attach(MIMEText(html_body, "html"))

    with smtplib.SMTP_SSL("smtp.gmail.com", 465) as server:
        server.login(GMAIL_USER, GMAIL_APP_PW)
        server.sendmail(GMAIL_USER, RECIPIENT, msg.as_string())

    print(f"Sent: {SUBJECT}")


# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    print("Fetching market data...")
    data = fetch_market_data()
    print(f"  BTC: ${data['price']:,.0f}  |  24h: {data['change_24h']:+.2f}%  |  F&G: {data['fg_value']} ({data['fg_label']})")

    print("Analysing...")
    analysis = analyse(data)

    print("Building report...")
    html = build_html(data, analysis)

    print("Sending email...")
    send(html)
